src/Lean/Elab/Coinductive.lean

```diff
@@ -0,0 +1,467 @@
+  between the user-facing coinductive predicates and their internal flat representations.
+-/
+private def generateEqLemmas (infos : Array InductiveVal) : MetaM Unit := do
+  let levels := infos[0]!.levelParams.map mkLevelParam
+  for info in infos do
+    let res ← forallTelescopeReducing info.type fun args _ => do
+      let params := args.take <| info.numParams - infos.size
+      let args := args.extract info.numParams
+
+      let lhs := mkConst (removeFunctorPostfix info.name) levels
+      let lhs := mkAppN lhs params
+      let lhs := mkAppN lhs args
+
+      let calls := infos.map fun info => mkAppN (mkConst (removeFunctorPostfix info.name) levels) params
+      let rhs := mkConst info.name levels
+      let rhs := mkAppN rhs (params ++ calls ++ args)
+
+      let goalType ← mkEq lhs rhs
+      let goal ← mkFreshExprMVar goalType
+
+      let goalMVarId := goal.mvarId!
+
+      let .some #[fixEq] ←  getEqnsFor? (removeFunctorPostfix info.name) | throwError "did not generate unfolding theorem"
+      let existentialEquiv := mkConst (info.name ++ `existential_equiv) levels
+
+      let mut fixEq := mkConst fixEq levels
+      fixEq := mkAppN fixEq params
+      for arg in args do
+        fixEq ← mkCongrFun fixEq arg
+
+      let newGoal ← rewriteGoalUsingEq goalMVarId existentialEquiv
+      newGoal.assign fixEq
+
+      let goal ← instantiateMVars goal
+      mkLambdaFVars (params ++ args) goal
+    trace[Elab.coinductive] "res: {res}"
+    addDecl <| .defnDecl (←mkDefinitionValInferringUnsafe
+      ((removeFunctorPostfix info.name) ++ `functor_unfold) info.levelParams (←inferType res) res .opaque)
+
+/--
+  Generates a constructor for a coinductive predicate that corresponds to constructors
+  in the original `InductiveView`.
+
+  The process:
+  1. Takes the flat inductive constructor type
+  2. Fills recursive call parameters with the actual coinductive predicates
+  3. Converts to existential form using the equivalence lemma
+  4. Applies the unrolling rule to get the final constructor form
+-/
+private def generateCoinductiveConstructor (infos : Array InductiveVal) (ctorSyntax : Syntax)
+    (numParams : Nat) (name : Name) (ctor : ConstructorVal) : TermElabM Unit := do
+  trace[Elab.coinductive] "Generating constructor: {removeFunctorPostfixInCtor ctor.name}"
+  let numPreds := infos.size
+  let predNames := infos.map fun val => removeFunctorPostfix val.name
+  let levelParams := infos[0]!.levelParams.map mkLevelParam
+  /-
+    We start by looking at the type of the constructor of the flat inductive and then by introducing
+    all its parameters to the scope.
+  -/
+  forallBoundedTelescope ctor.type (numParams + numPreds) fun args body => do
+    /-
+      The first `numParams` many items of `args` are parameters from the original definition,
+      while the remaining ones are free variables that correspond to recursive calls.
+    -/
+    let params := args.take numParams
+    let predFVars := args.extract numParams
```

Comment:
Just style, but I find
```suggestion
    let params := args[:numParams]
    let predFVars := args[numParams:]
```
easier to parse and  understand. Especially with the less telling name `extract`. But no big deal either way, just in case you didn't consider this syntax.